[MSE] Add MediaSourceHandle object and update to MSE's latest spec (including MSE in worker)

jyavenard opened the pull request:
#### 7db823fea8a1af033195fe8ddcaad6d7fae675b5
<pre>
[MSE] Add MediaSourceHandle object and update to MSE&apos;s latest spec (including MSE in worker)
<a href="https://bugs.webkit.org/show_bug.cgi?id=269461">https://bugs.webkit.org/show_bug.cgi?id=269461</a>
<a href="https://rdar.apple.com/problem/123006531">rdar://problem/123006531</a>

Reviewed by Youenn Fablet.

The MSE spec now defines that a MediaSource can be constructed in a DedicatedWorker
and attached to a HTMLMediaElement running in the main thread through a MediaSourceHandle
object.
To determine if a MediaSource can be constructed in a DedicatedWorker a new attribute
has been added.
This commit adds support for all those latest functionalities and is placed behind a MediaSourceInWorker
pref currently off by default.

Some architectural explanations:

The MediaSource and SourceBuffer were ThreadSafeRefCounted objects which was
problematic as we couldn&apos;t guarantee which thread those objets would be destructed on.
To solve this, we make MediaSource and SourceBuffer be RefCounted instead and
use a ThreadSafeRefCounted {MediaSource|SourceBuffer}PrivateClient each providing
a function dispatcher used for the {MediaSource|SourceBuffer}Private to communicate
with their client.
The function dispatcher guarantees that:
1- The function will always be run on the MediaSource&apos;s thread.
2- That the MediaSource hasn&apos;t been destructed in-between calls
3- That the MediaSource&apos;s readyState is not &quot;Closed&quot;.

As such, {MediaSource|SourceBuffer} methods are only ever called on the correct thread,
it is no longer the responsibility of either the MediaSource or the SourceBuffer to
run on the correct thread, so we can remove the now redundant `assertIsCurrent(m_dispatcher)`
and the m_dispatcher member.
Guarantees 2 and 3 allows to remove the checks made in each method, allowing to simplify the code.

One of the existing complexity, was that the media element directly interacted with the MediaSource
through both sync and asynchronous methods which is no longer a possibility if the MediaSource
now exists in the DedicatedWorker.
We introduce the MediaSourceInterfaceProxy virtual interface so that the media element
can continue to interact with the MediaSource with the minimum amount of changes.
Two implementation are provided.
MediaSourceInterfaceMainThread, which performs a simple, passthrough connection
and the MediaSourceInterfaceWorker which is constructed from the MediaSourceHandle.

The MediaSourceHandle implementation contains the minimum information so that the
MediaSourceInterfaceProxy can be established.

While asynchronous calls can be easily achieved with the MediaSourceHandle&apos;s
function dispatcher, synchronous calls can&apos;t (you can&apos;t dispatch synchronously
to a worker)
As such, all information needed to be synchronously retrieved
by the HTMLMediaElement from the MediaSource are now moved to the MediaSourcePrivate
instead, this includes:
- readyState
- duration
- buffered range
- seekable range
- live seekable range.

In order to avoid duplicating such handling with all MediaSourcePrivate implementations:
(AVF, Mock, GStreamer, GPU) we place it in the MediaSourcePrivate base class.

Various small changes had to be performed for things to operate when a worker is in use.
- Tracks validity where being checked using main thread&apos;s only AtomString, we change the
code use litteral strings for now.
- Some calls from the MediaSource (worker) are made directly to the MediaPlayer (which
lives in the main thread), rather than making the code thread-safe, we instead
dispatch synchronously to the main thread.

Present issues:
- The handling of Tracks and TrackList in the worker is currently broken as the
MSE spec do not cater yet for necessary changes (such as {Audio|Video|Text}Track
objects are currently only exposed to the `Document` (see <a href="https://github.com/w3c/media-source/issues/280)">https://github.com/w3c/media-source/issues/280)</a>
- Media Logging in the Worker is currently disabled.
- There appear to still be some unnecessary hops via the main thread while a MediaSource is in
operation in the work, which results in some janks still occurring when the main thread is
under heavy load.
- When in the main thread, it is possible via dynamic preferences to limit the codec&apos;s supported
configuration. As Settings aren&apos;t available in the Worker, this functionality is disabled.

MediaSource in a Worker functionality is currently turn off by default.
Tests will be added at a later stage.

No changes in observable behaviour, currently covered by extensive existing tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediasource/DOMURLMediaSource.cpp:
(WebCore::DOMURLMediaSource::createObjectURL):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::setStreaming):
(WebCore::ManagedMediaSource::ensurePrefsRead):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
(WebCore::ManagedMediaSource::streamingTimerFired):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setRegistry):
(WebCore::MediaSource::MediaSource):
(WebCore::MediaSource::logger):
(WebCore::MediaSource::didLogMessage):
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::currentTime const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::waitForTarget):
(WebCore::MediaSource::completeSeek):
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::setLiveSeekableRange):
(WebCore::MediaSource::clearLiveSeekableRange):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDuration):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::setReadyState):
(WebCore::MediaSource::endOfStream):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::addSourceBuffer):
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::isTypeSupported):
(WebCore::MediaSource::isOpen const):
(WebCore::MediaSource::isClosed const):
(WebCore::MediaSource::isEnded const):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::attachToElement):
(WebCore::MediaSource::openIfInEndedState):
(WebCore::MediaSource::openIfDeferredOpen):
(WebCore::MediaSource::stop):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::activeRanges const):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
(WebCore::MediaSource::ensureWeakOnHTMLMediaElementContext const):
(WebCore::MediaSource::sourceBufferBufferedChanged):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::failedToCreateRenderer):
(WebCore::MediaSource::sourceBufferReceivedFirstInitializationSegmentChanged):
(WebCore::MediaSource::sourceBufferActiveTrackFlagChanged):
(WebCore::MediaSource::setMediaPlayerReadyState):
(WebCore::MediaSource::incrementDroppedFrameCount):
(WebCore::MediaSource::addAudioTrackToElement):
(WebCore::MediaSource::addTextTrackToElement):
(WebCore::MediaSource::addVideoTrackToElement):
(WebCore::MediaSource::addAudioTrackMirrorToElement):
(WebCore::MediaSource::addTextTrackMirrorToElement):
(WebCore::MediaSource::addVideoTrackMirrorToElement):
(WebCore::MediaSource::memoryPressure):
(WebCore::MediaSource::client const):
(WebCore::MediaSource::enabledForContext):
(WebCore::MediaSource::handle):
(WebCore::MediaSource::canConstructInDedicatedWorker):
(WebCore::MediaSource::settings const): Deleted.
(WebCore::MediaSource::ensureWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::promisedWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.idl:
* Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp: Added.
(WebCore::MediaSourceHandle::create):
(WebCore::MediaSourceHandle::MediaSourceHandle):
(WebCore::MediaSourceHandle::canDetach const):
(WebCore::MediaSourceHandle::setHasEverBeenAssignedAsSrcObject):
(WebCore::MediaSourceHandle::hasEverBeenAssignedAsSrcObject const):
(WebCore::MediaSourceHandle::isManaged const):
(WebCore::MediaSourceHandle::ensureOnDispatcher const):
(WebCore::MediaSourceHandle::detach):
(WebCore::MediaSourceHandle::mediaSourcePrivateClient const):
(WebCore::MediaSourceHandle::mediaSourceDidOpen):
(WebCore::MediaSourceHandle::mediaSourcePrivate const):
* Source/WebCore/Modules/mediasource/MediaSourceHandle.h: Added.
(WebCore::MediaSourceHandle::isDetached const):
(WebCore::MediaSourceHandle::setDetached):
* Source/WebCore/Modules/mediasource/MediaSourceHandle.idl: Copied from Source/WebCore/Modules/mediasource/ManagedMediaSource.idl.
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.cpp: Added.
(WebCore::MediaSourceInterfaceMainThread::MediaSourceInterfaceMainThread):
(WebCore::MediaSourceInterfaceMainThread::client const):
(WebCore::MediaSourceInterfaceMainThread::monitorSourceBuffers):
(WebCore::MediaSourceInterfaceMainThread::isClosed const):
(WebCore::MediaSourceInterfaceMainThread::duration const):
(WebCore::MediaSourceInterfaceMainThread::buffered const):
(WebCore::MediaSourceInterfaceMainThread::seekable const):
(WebCore::MediaSourceInterfaceMainThread::isStreamingContent const):
(WebCore::MediaSourceInterfaceMainThread::attachToElement):
(WebCore::MediaSourceInterfaceMainThread::detachFromElement):
(WebCore::MediaSourceInterfaceMainThread::openIfDeferredOpen):
(WebCore::MediaSourceInterfaceMainThread::isManaged const):
(WebCore::MediaSourceInterfaceMainThread::setAsSrcObject):
(WebCore::MediaSourceInterfaceMainThread::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
(WebCore::MediaSourceInterfaceMainThread::create):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp: Added.
(WebCore::MediaSourceInterfaceWorker::MediaSourceInterfaceWorker):
(WebCore::MediaSourceInterfaceWorker::client const):
(WebCore::MediaSourceInterfaceWorker::monitorSourceBuffers):
(WebCore::MediaSourceInterfaceWorker::isClosed const):
(WebCore::MediaSourceInterfaceWorker::duration const):
(WebCore::MediaSourceInterfaceWorker::buffered const):
(WebCore::MediaSourceInterfaceWorker::seekable const):
(WebCore::MediaSourceInterfaceWorker::isStreamingContent const):
(WebCore::MediaSourceInterfaceWorker::attachToElement):
(WebCore::MediaSourceInterfaceWorker::detachFromElement):
(WebCore::MediaSourceInterfaceWorker::openIfDeferredOpen):
(WebCore::MediaSourceInterfaceWorker::isManaged const):
(WebCore::MediaSourceInterfaceWorker::setAsSrcObject):
(WebCore::MediaSourceInterfaceWorker::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
(WebCore::MediaSourceInterfaceWorker::create):
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::SourceBuffer):
(WebCore::m_logIdentifier):
(WebCore::SourceBuffer::~SourceBuffer):
(WebCore::SourceBuffer::abort):
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::abortIfUpdating):
(WebCore::SourceBuffer::removedFromMediaSource):
(WebCore::SourceBuffer::seekToTime):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveRenderingError):
(WebCore::SourceBuffer::maximumBufferSize const):
(WebCore::SourceBuffer::setActive):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::validateInitializationSegment):
(WebCore::SourceBuffer::sourceBufferPrivateDurationChanged):
(WebCore::SourceBuffer::sourceBufferPrivateHighestPresentationTimestampChanged):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::bufferedSamplesForTrackId):
(WebCore::SourceBuffer::enqueuedSamplesForTrackID):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::enabledForContext):
(WebCore::SourceBuffer::settings const): Deleted.
(WebCore::SourceBuffer::ensureWeakOnDispatcher const): Deleted.
(WebCore::SourceBuffer::promisedWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.idl:
* Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::name):
(WebCore::isTypeExposedToGlobalObject):
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpMediaSourceHandle):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::takeDetachedMediaSourceHandles):
(WebCore::CloneDeserializer::readMediaSourceHandle):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::validateSerializedResult):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::canDetachMediaSourceHandles):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::loadResource):
(WebCore::HTMLMediaElement::hasManagedMediaSource const):
(WebCore::HTMLMediaElement::detachMediaSource):
(WebCore::HTMLMediaElement::mediaState const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::isValidKind const):
(WebCore::AudioTrack::updateKindFromPrivate):
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::isValidKind const):
(WebCore::VideoTrack::updateKindFromPrivate):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::applicationOctetStream):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::MediaSourcePrivate):
(WebCore::MediaSourcePrivate::seekable const):
(WebCore::MediaSourcePrivate::setLiveSeekableRange):
(WebCore::MediaSourcePrivate::clearLiveSeekableRange):
(WebCore::MediaSourcePrivate::liveSeekableRange const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
(WebCore::MediaSourcePrivateClient::logger const):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setClient):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::MediaSourcePrivateRemote):
(WebKit::MediaSourcePrivateRemote::mediaPlayerReadyState const):
(WebKit::MediaSourcePrivateRemote::setMediaPlayerReadyState):
(WebKit::MediaSourcePrivateRemote::MessageReceiver::mediaSourcePrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/274853@main">https://commits.webkit.org/274853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66cdd4288ef31fbe806e5e31e334f80e20e699f7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40165 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19177 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42710 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42472 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22100 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16506 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16161 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34690 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13963 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35642 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43988 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/33618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36442 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35971 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39791 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12288 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16599 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46800 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9014 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16648 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9631 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -93,10 +102,95 @@ String convertEnumerationToString(MediaSourcePrivate::EndOfStreamStatus enumerat
+    }
+
+private:
+    MediaSourceClientImpl(MediaSource& parent)
```
explicit

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -93,10 +102,95 @@ String convertEnumerationToString(MediaSourcePrivate::EndOfStreamStatus enumerat
+public:
+    static Ref<MediaSourceClientImpl> create(MediaSource& parent) { return adoptRef(*new MediaSourceClientImpl(parent)); }
+
+    void ensureWeakOnDispatcher(Function<void()>&& function, bool forceRun = false) const
```
Does it need to be public?

jyavenard replied:
would have to make it friend class for MediaSourceHandle to access it.
Also, this class is internal and private to MediaSource, I actually considered to making it a struct for ease.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -93,10 +102,95 @@ String convertEnumerationToString(MediaSourcePrivate::EndOfStreamStatus enumerat
+
+    void setPrivateAndOpen(Ref<MediaSourcePrivate>&& mediaSourcePrivate) final
+    {
+        ensureWeakOnDispatcher([parent = m_parent, mediaSourcePrivate = WTFMove(mediaSourcePrivate)]() mutable {
```
Maybe ensureWeakOnDispatcher should take a Function<void(MediaSource&)>&& instead of having to capture the weak pointer twice.

jyavenard replied:
great idea.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -839,8 +899,11 @@ ExceptionOr<void> MediaSource::removeSourceBuffer(SourceBuffer& buffer)
-                    m_mediaElement->removeAudioTrack(track);
+                ensureOnMainThread([weakMediaElement = m_mediaElement, track = Ref { track }]() mutable {
+                    // FIXME: Need to send a mirror when we are in a worker.
+                    if (weakMediaElement)
```
I would tend to RefPtr all the weakMediaElements here and below.

jyavenard replied:
they are made in RefPtr except where the operation is a one-liner

jyavenard replied:
did them all

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -850,8 +913,12 @@ ExceptionOr<void> MediaSource::removeSourceBuffer(SourceBuffer& buffer)
-            if (removedEnabledAudioTrack && m_mediaElement)
-                m_mediaElement->ensureAudioTracks().scheduleChangeEvent();
+            if (removedEnabledAudioTrack) {
+                ensureOnMainThread([weakMediaElement = m_mediaElement] {
```
Since this is not a sync call, there is no guarantee that the media element is still attached to this media source.
We might want to add a check for that whenever we are doing this pattern.

jyavenard replied:
hmmm, we test the value of weakMediaElement here, if we the MediaSource is on the main thread, it is a sync call in practice

youennf reviewed (commented):
Some early comments, need to stop for now.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -140,18 +251,28 @@ void MediaSource::setPrivateAndOpen(Ref<MediaSourcePrivate>&& mediaSourcePrivate
+#if ENABLE(MEDIA_SOURCE_IN_WORKERS)
+    if (m_handle) {
+        m_handle->mediaSourceDidOpen(*m_private);
+        for (auto& weakHandle : m_transferredHandles) {
```
This might be too late for this patch but I do not think there is a need for m_transferredHandles.
First, there is only one active handle at any given time, so the last wins.
In practice, it is probably better to only have to do `m_handle->mediaSourceDidOpen(*m_private);`.
Each handle could for instance have a Ref to an object where would be stored m_private.
When transferring, you pass along the Ref to that object.

Also, it seems what we are doing is that we get the private in main thread and we hop to worker thread to set it on the MediaSourceHandle that is mostly a main thread object.
A future refactoring could probably make this all main thread to remove any potential race.

jyavenard replied:
removed, used a ThreadSafeShared member shared between instances instead

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp:
```diff
@@ -0,0 +1,131 @@
+
+RefPtr<MediaSourcePrivate> MediaSourceHandle::mediaSourcePrivate() const
+{
+    Locker locker { m_lock };
```
I would guess that this one is only called from main thread?
If so, it would be nice if we could set m_private in main thread as well.
Could be done as a follow-up.

jyavenard replied:
I would need to use the MediaSourceInterfaceProxy interface to pass the MediaSourcePrivate to the MediaSource rather than from the MediaSourcePrivateClient.

I'll take this as a follow-up

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1360,18 +1416,14 @@ WTFLogChannel& MediaSource::logChannel() const
 #endif
 
-void MediaSource::failedToCreateRenderer(RendererType type)
+void MediaSource::failedToCreateRenderer(MediaSourcePrivateClient::RendererType type)
```
You could be doing using RenderType = MediaSourcePrivateClient::RendererType in MediaSource class definition

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -126,6 +219,24 @@ MediaSource::~MediaSource()
+#if !RELEASE_LOG_DISABLED
+Ref<Logger> MediaSource::logger(ScriptExecutionContext& context)
+{
+    if (RefPtr document = dynamicDowncast<Document>(context))
+        return document->logger();
```
Hm, this is something I didn't consider: moving to a worker means losing logging. :-/

jyavenard replied:
fixed

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -126,6 +219,24 @@ MediaSource::~MediaSource()
+}
+
+void MediaSource::didLogMessage(const WTFLogChannel&, WTFLogLevel, Vector<JSONLogValue>&&)
+{
+    // FIXME: Add logging for when MediaSource is running in worker.
+}
```
I think you can do something like: 
```
scriptExecutionContext()->postTaskToResponsibleDocument([channel, level, log] (Document& document) {
    document->didLogMessage(channel, level, WTFMove(log));
});
```

jyavenard replied:
didLogMessage is a private member of Document

youennf replied:
Document logger can probably be used for the MediaSource as it seems to be thread safe. This seems best be done as a follow-up.

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -269,80 +382,47 @@ void MediaSource::completeSeek()
-    }))->whenSettled(m_dispatcher, [time = seekTarget.time, protectedThis = Ref { *this }] (auto&& results) mutable {
+    })), TaskSource::MediaElement, [producer = WTFMove(producer), weakThis = WeakPtr { *this }, this, time = seekTarget.time](auto&& results) {
+        if (!weakThis || isClosed())
+            return;
```
Don't you need to wrap `weakThis` in a `protectedThis` here? Otherwise there's nothing keeping `this` alive during the run of this method.

jyavenard replied:
There's no external call that could make the MEdiaSource be deleted; and while the task is running, it can't be deleted. The MediaSource is only ever accessed on the context's dispatcher

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -739,15 +800,16 @@ ExceptionOr<Ref<SourceBuffer>> MediaSource::addSourceBuffer(const String& type)
 
+    // 2. If type contains a MIME type that is not supported ..., then throw a
+    // NotSupportedError exception and abort these steps.
+    Vector<ContentType> mediaContentTypesRequiringHardwareSupport;
+    if (RefPtr document = dynamicDowncast<Document>(context))
+        mediaContentTypesRequiringHardwareSupport.appendVector(document->settings().mediaContentTypesRequiringHardwareSupport());
+
```
Can you eliminate the `document` requirement here by using `scriptExecutionContext()->settingsValues()`?

jyavenard replied:
SettingsValues doesn't support Vector.

jyavenard replied:
I'll take this in a follow-up fix.

jyavenard replied:
[bug ](https://bugs.webkit.org/show_bug.cgi?id=269567)

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -988,17 +1069,19 @@ bool MediaSource::isTypeSupported(ScriptExecutionContext& context, const String&
     if (RefPtr document = dynamicDowncast<Document>(context)) {
-        auto& settings = document->settings();
-        if (!contentTypeMeetsContainerAndCodecTypeRequirements(contentType, settings.allowedMediaContainerTypes(), settings.allowedMediaCodecTypes()))
+        if (!contentTypeMeetsContainerAndCodecTypeRequirements(contentType, document->settings().allowedMediaContainerTypes(), document->settings().allowedMediaCodecTypes()))
             return false;
```
Ditto.

jyavenard replied:
Same as above, those preferences are only found in `Settings` not in `SettingsValues` (a Settings class contain a SettingsValues class for storing the numerical value.

`Settings` live in the Document; not a Worker.

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -988,17 +1069,19 @@ bool MediaSource::isTypeSupported(ScriptExecutionContext& context, const String&
-        parameters.allowedMediaVideoCodecIDs = settings.allowedMediaVideoCodecIDs();
-        parameters.allowedMediaAudioCodecIDs = settings.allowedMediaAudioCodecIDs();
-        parameters.allowedMediaCaptionFormatTypes = settings.allowedMediaCaptionFormatTypes();
+        parameters.allowedMediaContainerTypes = document->settings().allowedMediaContainerTypes();
+        parameters.allowedMediaVideoCodecIDs = document->settings().allowedMediaVideoCodecIDs();
+        parameters.allowedMediaAudioCodecIDs = document->settings().allowedMediaAudioCodecIDs();
+        parameters.allowedMediaCaptionFormatTypes = document->settings().allowedMediaCaptionFormatTypes();
```
Ditto.

jyavenard replied:
ditto ditto :)

jernoble commented on Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
```diff
@@ -170,6 +172,75 @@ const PlatformTimeRanges& MediaSourcePrivate::buffered() const
+
+const PlatformTimeRanges& MediaSourcePrivate::liveSeekableRange() const
+{
+    Locker locker { m_lock };
+
+    return m_liveSeekable;
```
This isn't a valid lock, since you're handing back a reference to the underlying object, which can be mutated on another thread. This would need to return a new PlatformTimeRanges object each time it's called.

However, it doesn't look like this method is used. Can it just be removed?

jyavenard replied:
doh, I don't know what I was thinking here. I've fixed access to m_buffered in a separate PR #24602

youennf commented on Source/WebCore/Modules/mediasource/DOMURLMediaSource.cpp:
```diff
@@ -41,9 +41,9 @@ namespace WebCore {
-    return DOMURL::createPublicURL(scriptExecutionContext, source);
+    if (is<Document>(scriptExecutionContext))
+        return DOMURL::createPublicURL(scriptExecutionContext, source);
+    return nullString();
```
Do we have test for that?

jyavenard replied:
It's one of the wpt test, I'll do so in https://bugs.webkit.org/show_bug.cgi?id=269564

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -93,10 +102,95 @@ String convertEnumerationToString(MediaSourcePrivate::EndOfStreamStatus enumerat
+
+    const Logger* logger() const
+    {
+        if (m_isMainThread)
```
It seems this is the only use.
Maybe we could simply make m_logger a RefPtr and check isMainThread once in constructor.

Also I do not see where we are using logger(), can we remove it?

jyavenard replied:
will remove.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -140,18 +252,21 @@ void MediaSource::setPrivateAndOpen(Ref<MediaSourcePrivate>&& mediaSourcePrivate
-        ensureOnMainThread([this, protectedThis = Ref { *this }] {
-            if (m_mediaElement)
-                m_mediaElement->mediaLoadingFailedFatally(MediaPlayer::NetworkState::NetworkError);
+        ensureWeakOnHTMLMediaElementContext([](HTMLMediaElement& mediaElement) {
```
auto would read well here.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1396,74 +1441,111 @@ void MediaSource::sourceBufferActiveTrackFlagChanged(bool activeTrackFlag)
+bool MediaSource::enabledForContext(ScriptExecutionContext& context)
+{
+    UNUSED_PARAM(context);
+#if ENABLE(GPU_PROCESS)
```
Why is it not `#if ENABLE(MEDIA_SOURCE_IN_WORKERS)`

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp:
```diff
@@ -0,0 +1,148 @@
+        , m_dispatcher(WTFMove(dispatcher))
+    {
+    }
+    mutable Lock m_lock;
```
I think structs usually do not have m_ names but plain names.
I would tend to make this a class though, given the locks and so on.

jyavenard replied:
I just wanted all members to be public.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp:
```diff
@@ -0,0 +1,148 @@
+    return adoptRef(*new MediaSourceHandle(mediaSource, WTFMove(dispatcher)));
+}
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(Ref<MediaSourceHandle>&& other)
```
This is difficult to read and follow.
I see why it is there but I feel like it would just be better to use the detached handle as a Ref<MediaSourceHandlePrivate> and create a MediaSourceHandle from Ref<MediaSourceHandlePrivate> directly.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp:
```diff
@@ -0,0 +1,148 @@
+
+Ref<DetachedMediaSourceHandle> MediaSourceHandle::detach()
+{
+    return adoptRef(*new MediaSourceHandle(*this));
```
It would read better if we would do something like:
```
m_detached = true;
return ...
```
See above where it seems we could just send a ref to the private.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp:
```diff
@@ -0,0 +1,148 @@
+
+void MediaSourceHandle::ensureOnDispatcher(MediaSourceHandle::TaskType&& task) const
+{
+    m_private->dispatch(WTFMove(task));
```
We could: ASSERT(!isDetached()) just for caution.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.h:
```diff
@@ -0,0 +1,82 @@
+
+    static Ref<MediaSourceHandle> create(MediaSource&, DispatcherType&&);
+    MediaSourceHandle(MediaSource&, DispatcherType&&);
+    MediaSourceHandle(MediaSourceHandle&);
```
explicit, but I would hope we could replace this one with Ref<SharedPrivate>&&.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceHandle.h:
```diff
@@ -0,0 +1,82 @@
+    void setDetached(bool value) { m_detached = value; }
+
+    bool m_detached { false };
+    Ref<SharedPrivate> m_private;
```
If using a RefPtr, we could remove m_detached.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h:
```diff
@@ -0,0 +1,60 @@
+public:
+    static Ref<MediaSourceInterfaceMainThread> create(Ref<MediaSource>&& mediaSource) { return adoptRef(*new MediaSourceInterfaceMainThread(WTFMove(mediaSource))); }
+
+    bool isMainThread() const final { return true; }
```
Do they all need to be public?

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebCore {
+
+class MediaSourceInterfaceMainThread : public MediaSourceInterfaceProxy {
```
final

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h:
```diff
@@ -0,0 +1,60 @@
+    void memoryPressure() final;
+
+private:
+    MediaSourceInterfaceMainThread(Ref<MediaSource>&&);
```
explicit plus maybe add an empty line below.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h:
```diff
@@ -0,0 +1,66 @@
+class MediaSourceInterfaceProxy
+    : public RefCounted<MediaSourceInterfaceProxy>
+    , public CanMakeWeakPtr<MediaSourceInterfaceProxy> {
+
```
Line not needed

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h:
```diff
@@ -0,0 +1,66 @@
+    virtual PlatformTimeRanges buffered() const = 0;
+    virtual Ref<TimeRanges> seekable() const = 0;
+    virtual bool isStreamingContent() const = 0;
+    virtual bool attachToElement(WeakPtr<HTMLMediaElement>&&) = 0;
```
Should probably be attachToElement(HTMLMediaElement&) here.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1064,22 +1130,24 @@ void MediaSource::sourceBufferDidChangeActiveState(SourceBuffer&, bool)
 }
 
-bool MediaSource::attachToElement(HTMLMediaElement& element)
+bool MediaSource::attachToElement(WeakPtr<HTMLMediaElement>&& element)
```
Maybe we should add two attachToElement, one with a HTMLMediaElement&, called by the main proxy, and one with a WeakPtr<>&& called by the worker proxy.

The second one would do nothing but set m_mediaElement, and would probably not even return a boolean.

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp:
```diff
@@ -0,0 +1,141 @@
+    return m_handle->isManaged();
+}
+
+void MediaSourceInterfaceWorker::setAsSrcObject(bool set)
```
s/set/isSet/

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h:
```diff
@@ -0,0 +1,63 @@
+public:
+    static Ref<MediaSourceInterfaceWorker> create(Ref<MediaSourceHandle>&& handle) { return adoptRef(*new MediaSourceInterfaceWorker(WTFMove(handle))); }
+
+    bool isMainThread() const final { return false; }
```
private?

jyavenard replied:
I've removed the method isMainThread

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h:
```diff
@@ -0,0 +1,63 @@
+    void memoryPressure() final;
+
+private:
+    MediaSourceInterfaceWorker(Ref<MediaSourceHandle>&&);
```
explicit, maybe an empty line below

youennf commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -75,6 +75,95 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(SourceBuffer);
+    explicit SourceBufferClientImpl(SourceBuffer& parent)
+        : m_parent(parent)
+        , m_identifier(parent.scriptExecutionContext()->identifier())
+        , m_isMainThread(isMainThread())
```
It does not seem to be much used, can we remove it?

youennf commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -85,35 +174,27 @@ Ref<SourceBuffer> SourceBuffer::create(Ref<SourceBufferPrivate>&& sourceBufferPr
-    , m_dispatcher(RunLoop::current())
 #if !RELEASE_LOG_DISABLED
-    , m_logger(m_private->sourceBufferLogger())
+    , m_logger(source.scriptExecutionContext()->isWorkletGlobalScope() ? source.logger() : m_private->sourceBufferLogger())
```
isWorkletGlobalScope?
This is probably wrong, is it isWorkerGlobalScope?

youennf commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -627,7 +718,7 @@ uint64_t SourceBuffer::maximumBufferSize() const
     const float bufferBudgetPercentageForAudio = .05;
 
-    size_t maximum = settings().maximumSourceBufferSize();
+    size_t maximum = scriptExecutionContext()->settingsValues().maximumSourceBufferSize;
```
Are we sure scriptExecutionContext() is always non nullptr?
If not, add a null check or store it in sourceBuffer constructor

youennf reviewed (approved):
Great work!
See comments below, most of them can be handled as follow-ups (like some previous comments not addressed here).

jyavenard committed:
[MSE] Add MediaSourceHandle object and update to MSE's latest spec (including MSE in worker)
https://bugs.webkit.org/show_bug.cgi?id=269461
rdar://problem/123006531

Reviewed by Youenn Fablet.

The MSE spec now defines that a MediaSource can be constructed in a DedicatedWorker
and attached to a HTMLMediaElement running in the main thread through a MediaSourceHandle
object.
To determine if a MediaSource can be constructed in a DedicatedWorker a new attribute
has been added.
This commit adds support for all those latest functionalities and is placed behind a MediaSourceInWorker
pref currently off by default.

Some architectural explanations:

The MediaSource and SourceBuffer were ThreadSafeRefCounted objects which was
problematic as we couldn't guarantee which thread those objets would be destructed on.
To solve this, we make MediaSource and SourceBuffer be RefCounted instead and
use a ThreadSafeRefCounted {MediaSource|SourceBuffer}PrivateClient each providing
a function dispatcher used for the {MediaSource|SourceBuffer}Private to communicate
with their client.
The function dispatcher guarantees that:
1- The function will always be run on the MediaSource's thread.
2- That the MediaSource hasn't been destructed in-between calls
3- That the MediaSource's readyState is not "Closed".

As such, {MediaSource|SourceBuffer} methods are only ever called on the correct thread,
it is no longer the responsibility of either the MediaSource or the SourceBuffer to
run on the correct thread, so we can remove the now redundant `assertIsCurrent(m_dispatcher)`
and the m_dispatcher member.
Guarantees 2 and 3 allows to remove the checks made in each method, allowing to simplify the code.

One of the existing complexity, was that the media element directly interacted with the MediaSource
through both sync and asynchronous methods which is no longer a possibility if the MediaSource
now exists in the DedicatedWorker.
We introduce the MediaSourceInterfaceProxy virtual interface so that the media element
can continue to interact with the MediaSource with the minimum amount of changes.
Two implementation are provided.
MediaSourceInterfaceMainThread, which performs a simple, passthrough connection
and the MediaSourceInterfaceWorker which is constructed from the MediaSourceHandle.

The MediaSourceHandle implementation contains the minimum information so that the
MediaSourceInterfaceProxy can be established.

While asynchronous calls can be easily achieved with the MediaSourceHandle's
function dispatcher, synchronous calls can't (you can't dispatch synchronously
to a worker)
As such, all information needed to be synchronously retrieved
by the HTMLMediaElement from the MediaSource are now moved to the MediaSourcePrivate
instead, this includes:
- readyState
- duration
- buffered range
- seekable range
- live seekable range.

In order to avoid duplicating such handling with all MediaSourcePrivate implementations:
(AVF, Mock, GStreamer, GPU) we place it in the MediaSourcePrivate base class.

Various small changes had to be performed for things to operate when a worker is in use.
- Tracks validity where being checked using main thread's only AtomString, we change the
code use litteral strings for now.
- Some calls from the MediaSource (worker) are made directly to the MediaPlayer (which
lives in the main thread), rather than making the code thread-safe, we instead
dispatch synchronously to the main thread.

Present issues:
- The handling of Tracks and TrackList in the worker is currently broken as the
MSE spec do not cater yet for necessary changes (such as {Audio|Video|Text}Track
objects are currently only exposed to the `Document` (see https://github.com/w3c/media-source/issues/280)
- Media Logging in the Worker is currently disabled.
- There appear to still be some unnecessary hops via the main thread while a MediaSource is in
operation in the work, which results in some janks still occurring when the main thread is
under heavy load.
- When in the main thread, it is possible via dynamic preferences to limit the codec's supported
configuration. As Settings aren't available in the Worker, this functionality is disabled.

MediaSource in a Worker functionality is currently turn off by default.
Tests will be added at a later stage.

No changes in observable behaviour, currently covered by extensive existing tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediasource/DOMURLMediaSource.cpp:
(WebCore::DOMURLMediaSource::createObjectURL):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::setStreaming):
(WebCore::ManagedMediaSource::ensurePrefsRead):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
(WebCore::ManagedMediaSource::streamingTimerFired):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setRegistry):
(WebCore::MediaSource::MediaSource):
(WebCore::MediaSource::logger):
(WebCore::MediaSource::didLogMessage):
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::currentTime const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::waitForTarget):
(WebCore::MediaSource::completeSeek):
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::setLiveSeekableRange):
(WebCore::MediaSource::clearLiveSeekableRange):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDuration):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::setReadyState):
(WebCore::MediaSource::endOfStream):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::addSourceBuffer):
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::isTypeSupported):
(WebCore::MediaSource::isOpen const):
(WebCore::MediaSource::isClosed const):
(WebCore::MediaSource::isEnded const):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::attachToElement):
(WebCore::MediaSource::openIfInEndedState):
(WebCore::MediaSource::openIfDeferredOpen):
(WebCore::MediaSource::stop):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::activeRanges const):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
(WebCore::MediaSource::ensureWeakOnHTMLMediaElementContext const):
(WebCore::MediaSource::sourceBufferBufferedChanged):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::failedToCreateRenderer):
(WebCore::MediaSource::sourceBufferReceivedFirstInitializationSegmentChanged):
(WebCore::MediaSource::sourceBufferActiveTrackFlagChanged):
(WebCore::MediaSource::setMediaPlayerReadyState):
(WebCore::MediaSource::incrementDroppedFrameCount):
(WebCore::MediaSource::addAudioTrackToElement):
(WebCore::MediaSource::addTextTrackToElement):
(WebCore::MediaSource::addVideoTrackToElement):
(WebCore::MediaSource::addAudioTrackMirrorToElement):
(WebCore::MediaSource::addTextTrackMirrorToElement):
(WebCore::MediaSource::addVideoTrackMirrorToElement):
(WebCore::MediaSource::memoryPressure):
(WebCore::MediaSource::client const):
(WebCore::MediaSource::enabledForContext):
(WebCore::MediaSource::handle):
(WebCore::MediaSource::canConstructInDedicatedWorker):
(WebCore::MediaSource::settings const): Deleted.
(WebCore::MediaSource::ensureWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::promisedWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.idl:
* Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp: Added.
(WebCore::MediaSourceHandle::create):
(WebCore::MediaSourceHandle::MediaSourceHandle):
(WebCore::MediaSourceHandle::canDetach const):
(WebCore::MediaSourceHandle::setHasEverBeenAssignedAsSrcObject):
(WebCore::MediaSourceHandle::hasEverBeenAssignedAsSrcObject const):
(WebCore::MediaSourceHandle::isManaged const):
(WebCore::MediaSourceHandle::ensureOnDispatcher const):
(WebCore::MediaSourceHandle::detach):
(WebCore::MediaSourceHandle::mediaSourcePrivateClient const):
(WebCore::MediaSourceHandle::mediaSourceDidOpen):
(WebCore::MediaSourceHandle::mediaSourcePrivate const):
* Source/WebCore/Modules/mediasource/MediaSourceHandle.h: Added.
(WebCore::MediaSourceHandle::isDetached const):
(WebCore::MediaSourceHandle::setDetached):
* Source/WebCore/Modules/mediasource/MediaSourceHandle.idl: Copied from Source/WebCore/Modules/mediasource/ManagedMediaSource.idl.
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.cpp: Added.
(WebCore::MediaSourceInterfaceMainThread::MediaSourceInterfaceMainThread):
(WebCore::MediaSourceInterfaceMainThread::client const):
(WebCore::MediaSourceInterfaceMainThread::monitorSourceBuffers):
(WebCore::MediaSourceInterfaceMainThread::isClosed const):
(WebCore::MediaSourceInterfaceMainThread::duration const):
(WebCore::MediaSourceInterfaceMainThread::buffered const):
(WebCore::MediaSourceInterfaceMainThread::seekable const):
(WebCore::MediaSourceInterfaceMainThread::isStreamingContent const):
(WebCore::MediaSourceInterfaceMainThread::attachToElement):
(WebCore::MediaSourceInterfaceMainThread::detachFromElement):
(WebCore::MediaSourceInterfaceMainThread::openIfDeferredOpen):
(WebCore::MediaSourceInterfaceMainThread::isManaged const):
(WebCore::MediaSourceInterfaceMainThread::setAsSrcObject):
(WebCore::MediaSourceInterfaceMainThread::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
(WebCore::MediaSourceInterfaceMainThread::create):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp: Added.
(WebCore::MediaSourceInterfaceWorker::MediaSourceInterfaceWorker):
(WebCore::MediaSourceInterfaceWorker::client const):
(WebCore::MediaSourceInterfaceWorker::monitorSourceBuffers):
(WebCore::MediaSourceInterfaceWorker::isClosed const):
(WebCore::MediaSourceInterfaceWorker::duration const):
(WebCore::MediaSourceInterfaceWorker::buffered const):
(WebCore::MediaSourceInterfaceWorker::seekable const):
(WebCore::MediaSourceInterfaceWorker::isStreamingContent const):
(WebCore::MediaSourceInterfaceWorker::attachToElement):
(WebCore::MediaSourceInterfaceWorker::detachFromElement):
(WebCore::MediaSourceInterfaceWorker::openIfDeferredOpen):
(WebCore::MediaSourceInterfaceWorker::isManaged const):
(WebCore::MediaSourceInterfaceWorker::setAsSrcObject):
(WebCore::MediaSourceInterfaceWorker::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h: Copied from Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp.
(WebCore::MediaSourceInterfaceWorker::create):
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::SourceBuffer):
(WebCore::m_logIdentifier):
(WebCore::SourceBuffer::~SourceBuffer):
(WebCore::SourceBuffer::abort):
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::abortIfUpdating):
(WebCore::SourceBuffer::removedFromMediaSource):
(WebCore::SourceBuffer::seekToTime):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveRenderingError):
(WebCore::SourceBuffer::maximumBufferSize const):
(WebCore::SourceBuffer::setActive):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::validateInitializationSegment):
(WebCore::SourceBuffer::sourceBufferPrivateDurationChanged):
(WebCore::SourceBuffer::sourceBufferPrivateHighestPresentationTimestampChanged):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::bufferedSamplesForTrackId):
(WebCore::SourceBuffer::enqueuedSamplesForTrackID):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::enabledForContext):
(WebCore::SourceBuffer::settings const): Deleted.
(WebCore::SourceBuffer::ensureWeakOnDispatcher const): Deleted.
(WebCore::SourceBuffer::promisedWeakOnDispatcher const): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.idl:
* Source/WebCore/Modules/pictureinpicture/PictureInPictureWindow.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::name):
(WebCore::isTypeExposedToGlobalObject):
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpMediaSourceHandle):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::takeDetachedMediaSourceHandles):
(WebCore::CloneDeserializer::readMediaSourceHandle):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::validateSerializedResult):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::canDetachMediaSourceHandles):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::loadResource):
(WebCore::HTMLMediaElement::hasManagedMediaSource const):
(WebCore::HTMLMediaElement::detachMediaSource):
(WebCore::HTMLMediaElement::mediaState const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::isValidKind const):
(WebCore::AudioTrack::updateKindFromPrivate):
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::isValidKind const):
(WebCore::VideoTrack::updateKindFromPrivate):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::applicationOctetStream):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::MediaSourcePrivate):
(WebCore::MediaSourcePrivate::seekable const):
(WebCore::MediaSourcePrivate::setLiveSeekableRange):
(WebCore::MediaSourcePrivate::clearLiveSeekableRange):
(WebCore::MediaSourcePrivate::liveSeekableRange const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
(WebCore::MediaSourcePrivateClient::logger const):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setClient):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::MediaSourcePrivateRemote):
(WebKit::MediaSourcePrivateRemote::mediaPlayerReadyState const):
(WebKit::MediaSourcePrivateRemote::setMediaPlayerReadyState):
(WebKit::MediaSourcePrivateRemote::MessageReceiver::mediaSourcePrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: https://commits.webkit.org/274853@main
